src/source/raster_tile_source.js

```diff
@@ -117,22 +120,25 @@ class RasterTileSource extends Evented implements Source {
                 delete (img: any).cacheControl;
                 delete (img: any).expires;
 
-                const context = this.map.painter.context;
-                const gl = context.gl;
-                tile.texture = this.map.painter.getTileTexture(img.width);
-                if (tile.texture) {
-                    tile.texture.update(img, { useMipmap: true });
-                } else {
-                    tile.texture = new Texture(context, img, gl.RGBA, { useMipmap: true });
-                    tile.texture.bind(gl.LINEAR, gl.CLAMP_TO_EDGE, gl.LINEAR_MIPMAP_NEAREST);
-
-                    if (context.extTextureFilterAnisotropic) {
-                        gl.texParameterf(gl.TEXTURE_2D, context.extTextureFilterAnisotropic.TEXTURE_MAX_ANISOTROPY_EXT, context.extTextureFilterAnisotropicMax);
+                function textureCallback (tile, img) {
+                    const context = this.map.painter.context;
+                    const gl = context.gl;
+                    tile.texture = this.map.painter.getTileTexture(img.width);
+                    if (tile.texture) {
+                        tile.texture.update(img, { useMipmap: true });
+                    } else {
+                        tile.texture = new Texture(context, img, gl.RGBA, { useMipmap: true });
+                        tile.texture.bind(gl.LINEAR, gl.CLAMP_TO_EDGE, gl.LINEAR_MIPMAP_NEAREST);
+
+                        if (context.extTextureFilterAnisotropic) {
+                            gl.texParameterf(gl.TEXTURE_2D, context.extTextureFilterAnisotropic.TEXTURE_MAX_ANISOTROPY_EXT, context.extTextureFilterAnisotropicMax);
+                        }
```

Comment:
Given that we're now deferring the action of uploading the tile, `abortTile` no longer cancels everything: There could be a situation where the request responds with a tile, we store in the queue, then the tile gets aborted/deleted. We now still hold on to the tile and upload it to the GPU, but it'll never be used and gets GCed once the callback/tile object is removed from the `textureQueue`. We should add a way to either remove requests from the `textureQueue` that reference this tile, or use another mechanism to prevent tiles from remaining alive longer than they need to.

Reply:
Is it enough to check that the tile pulled from `textureQueue` doesn't have the `aborted` property before uploading to the GPU? Or would that be unreliable and not cover all potential cases?

Reply:
That works; another option would be to remove the object from the queue when the tile is aborted.

Reply:
There's similar logic here if that helps: https://github.com/mapbox/mapbox-gl-js/blob/master/src/util/ajax.js#L196-L197